pkg/collector/corechecks/cluster/ksm/kubernetes_state.go

```diff
@@ -824,10 +827,13 @@ func (k *KSMCheck) sendTelemetry(s sender.Sender) {
 	// reset the cache for the next check run
 	defer k.telemetry.reset()
 
-	s.Gauge(ksmMetricPrefix+"telemetry.metrics.count.total", float64(k.telemetry.getTotal()), "", k.instance.Tags)
-	s.Gauge(ksmMetricPrefix+"telemetry.unknown_metrics.count", float64(k.telemetry.getUnknown()), "", k.instance.Tags) // useful to track metrics that aren't mapped to DD metrics
+	s.Gauge(ksmMetricPrefix+"telemetry.metrics.count.total", float64(k.telemetry.getTotal()), "", util.CopyArray(k.globalTags))
+	s.Gauge(ksmMetricPrefix+"telemetry.unknown_metrics.count", float64(k.telemetry.getUnknown()), "", util.CopyArray(k.globalTags))
 	for resource, count := range k.telemetry.getResourcesCount() {
-		s.Gauge(ksmMetricPrefix+"telemetry.metrics.count", float64(count), "", append(k.instance.Tags, "resource_name:"+resource))
+		tags := make([]string, len(k.globalTags)+1)
+		copy(tags, k.globalTags)
+		tags[len(k.globalTags)] = "resource_name:" + resource
+		s.Gauge(ksmMetricPrefix+"telemetry.metrics.count", float64(count), "", tags)
```

Comment:
I understood that `s.Gauge(…)` can potentially alter its `tags` fourth parameter but I feel that invoking several times `s.Gauge(…)` sequentially with the same `tags` parameter is a “classic” pattern.

Ex.: https://github.com/DataDog/datadog-agent/blob/e6670a877e73598a112b52db23eeea9803c18aa7/pkg/collector/corechecks/system/disk/iostats_nix.go#L163-L170

So, why would we need a `copy` at each call here ?

Reply:
> It seems that this PR is systematically replacing the use of k.instance.Tags by k.globalTags everywhere.
Why does it make any difference ?
It looks like it just changes the name of the variable. Or are there some places where we still use k.instance.Tags ?

Before this PR, `k.instance.Tags` would contain all the tags in the check's configuration and the global tags. Since we removed the former, I think it's better to call them `globalTags` for readability. Normally we shouldn't need `k.instance.Tags` because they are already attached in the sender.


> I understood that s.Gauge(…) can potentially alter its tags fourth parameter but I feel that invoking several times s.Gauge(…) sequentially with the same tags parameter is a “classic” pattern.

They use a safe `Sender` which performs a `copy` on the `tags` slice when it's called. In KSM we use a RawSender mutating the array in place and in another thread so we can have a data race and potentially an unknown behavior.

Reply:
Thank you for all those precision !

I would have two questions then:

First, if the “global tags” are supposed to be attached to every single metric, exactly like the check’s configuration tags, why aren’t they both attached to the sender ?

It should be as simple as doing something like:
```go
s, err := k.GetSender()
if err != nil {
	log.Errorf("failed to retrieve a sender for check %s: %s", string(c.ID()), err)
}
s.SetCheckCustomTags(k.instance.Tags)
```
at the end of `func (k *KSMCheck) initTags()`.

And once done, we could simplify a lot of code by removing `k.instance.Tags` (in `main` branch) or `k.globalTags` (in this PR branch) at every call to `s.Gauge(…)` (and the equivalent for other metric types).

The “telemetry” calls to `s.Gauge(…)` could pass `nil` as their last `tags` parameter and it would completely remove the issue.


Second,
if there’s a `RawSender` that mutates its last `tags` parameter and a `safeSender` that does a copy of it in order to be safe in concurrent contexts,
and if, for the “telemetry” calls to `Gauge(…)`, we violate the condition of usage of `RawSender` by reusing the `tags` parameter after the `Gauge(…)` invocation
https://github.com/DataDog/datadog-agent/blob/fcdcde1d79eab5cceeff73ec3c0b1f6ea0351f8e/pkg/collector/corechecks/cluster/ksm/kubernetes_state.go#L463-L467

then, doesn’t it just mean that, for the “telemetry” data, we just need a `safeSender` instead of a `RawSender` ?
I think it would be better to use the more appropriate API rather than building workarounds around an inappropriate API.